src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -0,0 +1,1282 @@
+
+                        case FrameType.Data:
+                            await ProcessDataFrame(frameHeader).ConfigureAwait(false);
+                            break;
+
+                        case FrameType.Settings:
+                            await ProcessSettingsFrame(frameHeader).ConfigureAwait(false);
+                            break;
+
+                        case FrameType.Priority:
+                            ProcessPriorityFrame(frameHeader);
+                            break;
+
+                        case FrameType.Ping:
+                            await ProcessPingFrame(frameHeader).ConfigureAwait(false);
+                            break;
+
+                        case FrameType.WindowUpdate:
+                            ProcessWindowUpdateFrame(frameHeader);
+                            break;
+
+                        case FrameType.RstStream:
+                            ProcessRstStreamFrame(frameHeader);
+                            break;
+
+                        case FrameType.GoAway:
+                            ProcessGoAwayFrame(frameHeader);
+                            break;
+
+                        case FrameType.PushPromise:     // Should not happen, since we disable this in our initial SETTINGS (TODO: We aren't currently, but we should)
+                        case FrameType.Continuation:    // Should only be received while processing headers in ProcessHeadersFrame
+                        default:
+                            throw new Http2ProtocolException(Http2ProtocolErrorCode.ProtocolError);
+                    }
+                }
+            }
+            catch (Exception)
+            {
+                AbortStreams(0);
+            }
+        }
+
+        // Note, this will return null for a streamId that's not in use.
+        // Callers must check for this and send a RST_STREAM or ignore as appropriate.
+        private Http2Stream GetStream(int streamId)
+        {
+            if (streamId <= 0)
+            {
+                throw new Http2ProtocolException(Http2ProtocolErrorCode.ProtocolError);
+            }
+
+            lock (_syncObject)
+            {
+                if (!_httpStreams.TryGetValue(streamId, out Http2Stream http2Stream))
+                {
+                    return null;
+                }
+
+                return http2Stream;
+            }
+        }
+
+        private async Task SendRstStreamFrameAsync(int streamId, Http2ProtocolErrorCode errorCode)
+        {
+            WriteFrameHeader(new FrameHeader(FrameHeader.RstStreamLength, FrameType.RstStream, FrameFlags.None, streamId));
+
+            _outgoingBuffer.AvailableSpan[0] = (byte)(((int)errorCode & 0xFF000000) >> 24);
+            _outgoingBuffer.AvailableSpan[1] = (byte)(((int)errorCode & 0x00FF0000) >> 16);
+            _outgoingBuffer.AvailableSpan[2] = (byte)(((int)errorCode & 0x0000FF00) >> 8);
+            _outgoingBuffer.AvailableSpan[3] = (byte)((int)errorCode & 0x000000FF);
```

Comment:
It'd be nice to call AvailableSpan once, store it into a local, and then fill its bytes.  Otherwise we're constructing a span for each byte.

Reply:
This deserves more structure like helpers where you say "write this int to this offset".

Reply:
Yeah, it should really be using BinaryPrimitives. I will fix, but probably in a subsequent PR.